### ISSUE TITLE
Recurrent freeze/deadlock on request.

### ISSUE THREAD

**Issue by mockthebear** [brunoos/luasec#128]
Allright. I have a telegram lua bot and its based on https.request.

Every X seconds it runs this code
`local success, code, headers, status = https.request{
    url = "https://api.telegram.org/bot" .. M.token .. "/" .. method,
    method = "POST",
    headers = {
      ["Content-Type"] =  "multipart/form-data; boundary=" .. boundary,
    	["Content-Length"] = string.len(body),
    },
    source = ltn12.source.string(body),
    sink = ltn12.sink.table(response),
  }`
I set some prints before and after that to make sure there is the problem. Even changed the tabe; to be made before then sending to  https.request

What i saw is, after many calls of  https.request... like 40+ it tend to get permanently stuck inside the call.
sometimes it runs for days, sometimes just a few hours it freezes. I tried to set the timeout to 10 seconds with no sucess.

The only way is killing the program.

Im running in raspberry pi model B+ but had the same issue with ubuntu 16.0.4

**Comment by brunoos** [comment]
Sorry but I cannot reproduce your environment to test. You should instrument http.lua and https.lua to give more information.

**Comment by mockthebear** [comment]
Hello brunoos. I think i figured out what is happening. 

Aparently, the connection were been closed or stood alive without send anything during the ssl handshake
The function freezed here:
https.lua line 93
`self.sock:dohandshake()`
Analyzing where it stuck

The last segment of code where it stuck were at ssl.c line 127
`case SSL_ERROR_WANT_READ:
err = socket_waitfd(&ssl->sock, WAITFD_R, tm);`

Aparently _tm_ has values of -1, -1, (clock start). No timeout set.

Because of that, socket_waitfd calls _poll_ without timeout and hangs forever.
I belive it were closed during this poll because even closing the connection it kept stuck.

I managed to fix that in a makeshift way adding tm->total = 10.0; so when that happens it just timeouts just fine. Maybe the ssl must set some timeout in this or the global operation?
